polynote-server/src/main/scala/polynote/server/SharedNotebook.scala

```
@@ -0,0 +1,285 @@
+      updateBuffer.add(newGlobalVersion, update)
+      applyUpdate(newGlobalVersion, subscriberId, update, versionPromise)
+  }.compile.drain.unsafeRunAsyncAndForget()
+
+  private def ensureKernel(): IO[Kernel[IO]] = kernelLock.acquire.bracket { _ =>
+    kernelRef.get.flatMap {
+      case None => kernelFactory.launchKernel(() => ref.get.map(_._2), statusUpdates).flatMap {
+        kernel => kernelRef.set(Some(kernel)).as(kernel)
+      }
+      case Some(kernel) => IO.pure(kernel)
+    }
+  }(_ => kernelLock.release)
+
+  // apply a versioned update from the queue, completing its version promise and updating the info about which
+  // versions the originating subscriber knows about
+  private def applyUpdate(newGlobalVersion: Int, subscriberId: Int, update: NotebookUpdate, versionPromise: Deferred[IO, GlobalVersion]) = ref.get.flatMap {
+    case (currentVer, notebook) =>
+
+      // TODO: remove this, just checking for now
+      assert(newGlobalVersion - 1 == currentVer, "Version is wrong!")
+
+      val doUpdate = update match {
+        case InsertCell(_, _, _, cell, after) => ref.set(newGlobalVersion -> notebook.insertCell(cell, after))
+        case DeleteCell(_, _, _, id)          => ref.set(newGlobalVersion -> notebook.deleteCell(id))
+        case UpdateCell(_, _, _, id, edits)   => ref.set(newGlobalVersion -> notebook.editCell(id, edits))
+        case UpdateConfig(_, _, _, config)    => ref.set(newGlobalVersion -> notebook.copy(config = Some(config)))
+        case SetCellLanguage(_, _, _, id, lang) => ref.set(newGlobalVersion -> notebook.updateCell(id)(_.copy(language = lang)))
+      }
+
+      for {
+        _  <- doUpdate
+        sub = subscribers.get(subscriberId)
+        _  <- if (sub != null) sub.setKnownVersions(update.globalVersion, update.localVersion) else IO.unit
+        _  <- updatesTopic.publish1(Some((newGlobalVersion, subscriberId, update)))
+        _  <- versionPromise.complete(newGlobalVersion).attempt
```

Is there a point in `.attempt` if we're returning an `IO[Unit]`? Will `.compile.drain.unsafeRunAsyncAndForget()` surface the error?

---

`attempt` gives you `IO[Either[Throwable, A]]`, so the error is not "thrown" and doesn't directly fail the IO. Since I'm throwing the (possibly failed) result away, the effect is that if `versionPromise.complete` fails then the failure will be ignored.  That would happen if somehow `versionPromise` was already completed.

I guess since I'm not expecting that to happen I should just let the failure bubble up. ¯\\\_(ツ)\_/¯

---

Ah I see, without the `attempt` if it errors we'd stop listening to the update queue. But maybe we should at least log the error :)